firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/UpdateAppClient.java

```diff
@@ -79,10 +109,185 @@ private void redirectToPlayForAabUpdate(String downloadUrl, Activity currentActi
     this.updateTask.updateProgress(updateState);
   }
 
+  public void updateApk(String downloadUrl, Activity currentActivity) {
+    getDownloadTask(downloadUrl)
+        .addOnSuccessListener(
+            file -> {
+              install(file.getPath(), currentActivity)
+                  .addOnSuccessListener(
+                      Void -> {
+                        UpdateState updateState =
+                            UpdateState.builder()
+                                .setApkTotalBytesToDownload(0)
+                                .setApkBytesDownloaded(file.length())
+                                .setUpdateStatus(UpdateStatus.INSTALLED)
+                                .build();
+                        updateTask.updateProgress(updateState);
+                        updateAppTaskCompletionSource.setResult(updateState);
+                      })
+                  .addOnFailureListener(
+                      e ->
+                          setUpdateAppTaskCompletionError(
+                              new FirebaseAppDistributionException(
+                                  Constants.ErrorMessages.NETWORK_ERROR,
+                                  FirebaseAppDistributionException.Status.INSTALLATION_FAILURE)));
+            })
+        .addOnFailureListener(
+            e ->
+                setUpdateAppErrorWithDefault(
+                    e,
+                    new FirebaseAppDistributionException(
+                        Constants.ErrorMessages.NETWORK_ERROR,
+                        FirebaseAppDistributionException.Status.DOWNLOAD_FAILURE)));
+  }
+
+  public Task<File> getDownloadTask(String downloadUrl) {
+    if (downloadTaskCompletionSource != null
+        && !downloadTaskCompletionSource.getTask().isComplete()) {
+      downloadCancellationTokenSource.cancel();
+    }
+
+    downloadCancellationTokenSource = new CancellationTokenSource();
+    downloadTaskCompletionSource =
+        new TaskCompletionSource<>(downloadCancellationTokenSource.getToken());
+
```

Comment:
maybe set download.pending on the progresslistener before starting the download?